test/helper/resource.go

```diff
@@ -128,27 +128,45 @@ func NewStatefulSet(namespace string, name string) *appsv1.StatefulSet {
 }
 
 // NewService will build a service object.
-func NewService(namespace string, name string) *corev1.Service {
-	return &corev1.Service{
-		TypeMeta: metav1.TypeMeta{
-			APIVersion: "v1",
-			Kind:       "Service",
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Namespace: namespace,
-			Name:      name,
-		},
-		Spec: corev1.ServiceSpec{
+func NewService(namespace string, name string, svcType corev1.ServiceType) *corev1.Service {
```

Comment:
How about this:

```diff
--- a/test/helper/resource.go
+++ b/test/helper/resource.go
@@ -128,7 +128,7 @@ func NewStatefulSet(namespace string, name string) *appsv1.StatefulSet {
 }

 // NewService will build a service object.
-func NewService(namespace string, name string) *corev1.Service {
+func NewService(namespace string, name string, svcType corev1.ServiceType) *corev1.Service {
        return &corev1.Service{
                TypeMeta: metav1.TypeMeta{
                        APIVersion: "v1",
@@ -147,7 +147,7 @@ func NewService(namespace string, name string) *corev1.Service {
                                        TargetPort: intstr.IntOrString{IntVal: 8080},
                                },
                        },
-                       Type: corev1.ServiceTypeClusterIP,
+                       Type: svcType,
                },
        }
```

Reply:
I have a test in my env.
A nodeport should be defined clearly,or isInvalid error will not happen 
 
![image](https://user-images.githubusercontent.com/17697155/169250813-b76ca57c-00bc-4fed-b85c-7e333ce737d6.png)

Reply:
A random port will be assigned to nodePort, refer to https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport

I think this error should not occur.

Reply:
> A random port will be assigned to nodePort, refer to https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport
> 
> I think this error should not occur.

yes,but the e2e_test is to test a fixed nodeport service cannot be promoted,so the nodeport args is needed.

Reply:
I probably understand what you mean, but I see that there is no testing for this in e2e.

Reply:
> but I see that there is no testing for this in e2e.

I added in karmactl_test.go

Reply:
New added e2e test run failed:
```log
 Unexpected error:
      <*errors.errorString | 0xc0005573c0>: {
          s: "failed to create resource \"/v1, Resource=services\"(karmadatest-wp7/service-jtg) in control plane: Service \"service-jtg\" is invalid: spec.ports[0].nodePort: Invalid value: 30002: provided port is already allocated",
      }
      failed to create resource "/v1, Resource=services"(karmadatest-wp7/service-jtg) in control plane: Service "service-jtg" is invalid: spec.ports[0].nodePort: Invalid value: 30002: provided port is already allocated
  occurred
  In [It] at: /home/runner/work/karmada/karmada/test/e2e/karmadactl_test.go:251
```